impl/maven-impl/src/main/java/org/apache/maven/impl/model/DefaultModelProcessor.java

```diff
@@ -97,30 +95,20 @@ public Path locateExistingPom(Path projectDirectory) {
     @Override
     public Model read(XmlReaderRequest request) throws IOException {
         Objects.requireNonNull(request, "source cannot be null");
-        Path pomFile = request.getPath();
-        if (pomFile != null) {
-            Path projectDirectory = pomFile.getParent();
-            List<ModelParserException> exceptions = new ArrayList<>();
-            for (ModelParser parser : modelParsers) {
-                try {
-                    Optional<Model> model =
-                            parser.locateAndParse(projectDirectory, Map.of(ModelParser.STRICT, request.isStrict()));
-                    if (model.isPresent()) {
-                        return model.get().withPomFile(pomFile);
-                    }
-                } catch (ModelParserException e) {
-                    exceptions.add(e);
-                }
-            }
-            try {
-                return doRead(request);
-            } catch (IOException e) {
-                exceptions.forEach(e::addSuppressed);
-                throw e;
-            }
-        } else {
-            return doRead(request);
-        }
+        return readOnSelfOrParent(request.getPath(), request);
+    }
+
+    private Model readOnSelfOrParent(Path pomFile, XmlReaderRequest request) throws IOException {
+        return pomFile == null
+                ? doRead(request)
+                : modelParsers.stream()
```

Comment:
extract concern into readParent() to hide impl. detail and give dedicated concern